docs/data/data-grid/recipes-editing/BulkEditingNoSnap.js

```diff
@@ -0,0 +1,185 @@
+              onClick={() => {
+                unsavedChangesRef.current.unsavedRows[id] = {
+                  ...row,
+                  _action: 'delete',
+                };
+                if (!unsavedChangesRef.current.rowsBeforeChange[id]) {
+                  unsavedChangesRef.current.rowsBeforeChange[id] = row;
+                }
+                setHasUnsavedRows(true);
+                apiRef.current.updateRows([row]); // to trigger row render
+              }}
+            />,
+          ];
+        },
+      },
+      ...data.columns,
+    ];
+  }, [data.columns, unsavedChangesRef, apiRef]);
+
+  const processRowUpdate = (newRow, oldRow) => {
+    const rowId = newRow.id;
+
+    unsavedChangesRef.current.unsavedRows[rowId] = newRow;
+    if (!unsavedChangesRef.current.rowsBeforeChange[rowId]) {
+      unsavedChangesRef.current.rowsBeforeChange[rowId] = oldRow;
+    }
+    setHasUnsavedRows(true);
+    return newRow;
+  };
+
+  const discardChanges = () => {
+    setHasUnsavedRows(false);
+    apiRef.current.updateRows(
+      Object.values(unsavedChangesRef.current.rowsBeforeChange),
+    );
+    unsavedChangesRef.current = {
+      unsavedRows: {},
+      rowsBeforeChange: {},
+    };
+  };
+
+  const saveChanges = async () => {
+    try {
+      // Persist updates in the database
+      setIsSaving(true);
+      await new Promise((resolve) => {
+        setTimeout(resolve, 1000);
+      });
+
+      setIsSaving(false);
+      const rowsToDelete = Object.values(
+        unsavedChangesRef.current.unsavedRows,
+      ).filter((row) => row._action === 'delete');
+      if (rowsToDelete.length > 0) {
+        apiRef.current.updateRows(rowsToDelete);
+      }
+
+      setHasUnsavedRows(false);
+      unsavedChangesRef.current = {
+        unsavedRows: {},
+        rowsBeforeChange: {},
+      };
+    } catch (error) {
+      setIsSaving(false);
+    }
+  };
+
+  return (
+    <div style={{ width: '100%' }}>
+      <div style={{ marginBottom: 8 }}>
+        <LoadingButton
+          disabled={!hasUnsavedRows}
+          loading={isSaving}
+          onClick={saveChanges}
+          startIcon={<SaveIcon />}
+          loadingPosition="start"
+        >
+          <span>Save</span>
+        </LoadingButton>
+        <Button
+          disabled={!hasUnsavedRows || isSaving}
+          onClick={discardChanges}
+          startIcon={<RestoreIcon />}
+        >
+          Discard all changes
+        </Button>
+      </div>
+      <div style={{ height: 400 }}>
+        <DataGridPremium
+          {...data}
+          columns={columns}
+          apiRef={apiRef}
+          disableRowSelectionOnClick
+          unstable_cellSelection
+          processRowUpdate={processRowUpdate}
+          experimentalFeatures={{ clipboardPaste: true }}
+          unstable_ignoreValueFormatterDuringExport
+          sx={{
+            '& .MuiDataGrid-row.row--removed': {
+              backgroundColor: 'rgba(255, 170, 170, 0.3)',
```

Comment:
The background color looks cool. 🤩 
Would it look more natural (closer to actual use-cases) to actually remove the deleted rows from the grid on delete click but store them separately in something like a `deletedRows` variable and revert them on clicking `Discard All Changes` button?

Reply:
> Would it look more natural (closer to actual use-cases)

It's actually based on a real use case from https://discord.com/channels/1131323012554174485/1131329358062178324/1146444366500614154

> to remove the deleted rows from the grid on delete click

I think the point here is to see the rows that you're going to delete before you actually do that 🙂